### DISCUSSION TITLE
能否更新包：当没有超链接（网址）时，自动屏蔽urldate域。

### DISCUSSION THREAD

**Discussion by earthmoon** [hushidong/biblatex-gb7714-2015#159] [General]
您好，

假如有一条文献为：
```
@article{liu2018c,
  title = {Partitioning Water Source and Sinking Process of a Groundwater-Dependent Desert Plant Community},
  author = {Liu, Ran and Wang, Yugang and Li, Congjuan and Ma, Jie and Li, Yan},
  date = {2018-09},
  journaltitle = {Plant and Soil},
  shortjournal = {Plant Soil},
  volume = {430},
  number = {1-2},
  pages = {73--85},
  issn = {0032-079X, 1573-5036},
  doi = {10.1007/s11104-018-3714-6},
  urldate = {2023-04-01},
  langid = {english}
}
```

当使用 _**biber**_ 编译时，会发现文末参考文献这一条有日期内容：

> [2023-04-01]

我对国标的理解是这样的：当文献有 _**url**_ 域时，_**urldate**_ 域应当被加上，此时比如期刊的缩写从 _**J**_ 也变为 _**J/OL**_ ；如果文献没有 _**url**_ 域时，此时比如期刊的缩写为 _**J**_ ，这时候不用加入 _**urldate**_ 域。目前只要  _**urldate**_ 域出现，则就会被加上，貌似没有分情况处理。

所以可不可以加一个判断条件，如果文末参考文献这一条显示有网址的超链接，比如 **_webofscience_** 的链接或者是从 **_doi_** 转的超链接，此时  _**urldate**_ 域加上；如果没有出现超链接也没有将 **_doi_** 转为超链接，此时文末参考文献不再显示  _**urldate**_ 域。

谢谢！

**Comment by hushidong** [comment]
这个问题是这样的，

若存在urldate 域，那么说明应该有url或doi，否则就是不完整的，所以此时应该补上。

若不想输出，则直接用url=false来去掉。

**Reply by earthmoon** [reply]
是这样的，当导入文献至文献管理器，有 **_url_** 或者 **_doi_** 或者两者都有是很正常的。有这些，文献信息才算更完整。利用`url=false`不合适，因为如果引用了网页，网页的 **_url_** 显然不可以不显示在文末的参考文献中。比如，这个问题的例子，其实是有 _**url**_ 的，但是学院说不用 _**url**_ ，所以我把 **_url_** 剔除了。学院也说都没有 **_doi_** ，所以我也通过`doi=false`让 **_doi_** 不出现在文末参考文献中。但是这样，_**urldate**_ 域却出现在文末参考文献中。所以 _**urldate**_  域是否出现最好可以依据当条参考文献在文末参考文献中是否有  _**url**_ 来判断最好。

**Reply by sikouhjw** [reply]
> 比如，这个问题的例子，其实是有 _**url**_ 的，但是学院说不用 _**url**_ ，所以我把 **_url_** 剔除了。

所以这里是学院的要求有问题，那就不要听学院的

**Reply by earthmoon** [reply]
> > 比如，这个问题的例子，其实是有 _**url**_ 的，但是学院说不用 _**url**_ ，所以我把 **_url_** 剔除了。
> 
> 所以这里是学院的要求有问题，那就不要听学院的

当时查看了武大的模板，模板里参考文献的样例确实没有 **_url_** ，里面都是 **[J]** ，而不是 **[J/OL]** 。

**Reply by sikouhjw** [reply]
> 当时查看了武大的模板，模板里参考文献的样例确实没有 **_url_** ，里面都是 **[J]** ，而不是 **[J/OL]** 。

没有给 `url` 的示例，不代表禁用 `url`，我建议：如果与国标差别不大的话，以国标为准

**Reply by earthmoon** [reply]
> > 当时查看了武大的模板，模板里参考文献的样例确实没有 **_url_** ，里面都是 **[J]** ，而不是 **[J/OL]** 。
> 
> 没有给 `url` 的示例，不代表禁用 `url`，我建议：如果与国标差别不大的话，以国标为准

而且还有个问题，**CNKI** 里面的中文文章，大部分都没有 **url**。

**Reply by sikouhjw** [reply]
> 而且还有个问题，**CNKI** 里面的中文文章，大部分都没有 **url**。

我的建议是：用谷歌学术导出，有 `url` 就有，没有就没有。

电子资源才必须有：获取和访问路径、数字对象唯标识。

**Reply by earthmoon** [reply]
> > 而且还有个问题，**CNKI** 里面的中文文章，大部分都没有 **url**。
> 
> 我的建议是：用谷歌学术导出，有 `url` 就有，没有就没有。
> 
> 电子资源才必须有：获取和访问路径、数字对象唯标识。

我感觉我刚刚的意思有歧义。**CNKI** 的导出的文章，按照国标，基本都属于电子析出文献，因此应该需要 **url**。但是，如果我们下载几篇 **CNKI** 的文章来看（查看期刊的中文排版），会发现绝大部分期刊都舍弃了 **url** 这个域。

**Reply by sikouhjw** [reply]
> 我感觉我刚刚的意思有歧义。**CNKI** 的导出的文章，按照国标，基本都属于电子析出文献，因此应该需要 **url**。

这是你的认识误区，CNKI 中的论文，都有对应的实体期刊。实体期刊可以提供 `url` 和 `doi`，也可以不提供。只有纯电子资源才必须要 `url`。[^1]

> 但是，如果我们下载几篇 **CNKI** 的文章来看（查看期刊的中文排版），会发现绝大部分期刊都舍弃了 **url** 这个域。

说明 CNKI 导出的 bib 不够好。

[^1]: https://github.com/zepinglee/gbt7714-bibtex-style/issues/87#issuecomment-949285876

**Reply by earthmoon** [reply]
原来只有纯电子资源才需要。实际只有纯电子的期刊很少。这说明普遍不应用 **url** 域也是正确的。武大模板没有 **url** 域也是正确的。

**Reply by hushidong** [reply]
url=false不影响online文献输出url的，所以你尽管用url=false选择就好了。

**Reply by earthmoon** [reply]
> url=false不影响online文献输出url的，所以你尽管用url=false选择就好了。

这么科学？牛。

**Reply by sikouhjw** [reply]
> 原来只有纯电子资源才需要。实际只有纯电子的期刊很少。

这是你的逻辑错误。纯电子资源『必须』要 `url`，有 `url` 的不一定是纯电子资源。实体期刊也可以有 `url`。

> 这说明普遍不应用 **url** 域也是正确的。武大模板没有 **url** 域也是正确的。

普遍没有 `url` 不代表什么，只代表这些 bib 还不够好，没有把信息提供完整。如果存在 `url`，最好提供，方便读者查阅。武大没有 `url` 不代表不可以使用，也不代表是正确的，这是你的逻辑错误。

**Reply by earthmoon** [reply]
> > 原来只有纯电子资源才需要。实际只有纯电子的期刊很少。
> 
> 这是你的逻辑错误。纯电子资源『必须』要 `url`，有 `url` 的不一定是纯电子资源。实体期刊也可以有 `url`。
> 
> > 这说明普遍不应用 **url** 域也是正确的。武大模板没有 **url** 域也是正确的。
> 
> 普遍没有 `url` 不代表什么，只代表这些 bib 还不够好，没有把信息提供完整。如果存在 `url`，最好提供，方便读者查阅。武大没有 `url` 不代表不可以使用，也不代表是正确的，这是你的逻辑错误。

是的。我一直觉得有超链接更好。只是武大模板、CNKI下载的文章文后的参考文献都没有超链接。所以被说的只有从众了。
